presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java

```diff
@@ -13,133 +13,266 @@
  */
 package com.facebook.presto.sql.analyzer;
 
+import com.facebook.presto.spi.SchemaTableName;
 import com.facebook.presto.sql.tree.AliasedRelation;
 import com.facebook.presto.sql.tree.ComparisonExpression;
 import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
 import com.facebook.presto.sql.tree.Join;
 import com.facebook.presto.sql.tree.JoinCriteria;
 import com.facebook.presto.sql.tree.JoinOn;
-import com.facebook.presto.sql.tree.JoinUsing;
 import com.facebook.presto.sql.tree.LogicalBinaryExpression;
-import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.SetOperation;
 import com.facebook.presto.sql.tree.Unnest;
+import com.google.common.collect.ImmutableList;
 
-import java.util.HashSet;
-import java.util.Set;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
 
+import static com.facebook.presto.metadata.MetadataUtil.toSchemaTableName;
+import static com.facebook.presto.spi.ConnectorMaterializedViewDefinition.TableColumn;
 import static com.facebook.presto.sql.analyzer.SemanticErrorCode.NOT_SUPPORTED;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
 
 // TODO: Add more cases https://github.com/prestodb/presto/issues/16032
 public class MaterializedViewPlanValidator
```

Comment:
(Major) this is a very weird class now. It entangles validation and extraction together. Kinda unclear to me what the relation between `query` and `analysis` is. Maybe let's chat in person to see how we could better design or even separate this class.

Reply:
Hi James, I separate the class into two, Validator and Extractor, and make Extractor inherit Validator to reuse the validation logic. In addition, both of the two classes are exposed only by a public static interface, so we control how the classes could be called.